### ISSUE TITLE
Retrieval of related model for a `HasMany` relation

### ISSUE THREAD

**Issue by b-admike** [strongloop/loopback-next#1372]
## Description
Follow up for #1342. Given `Customer` and `Order` models, and a customer that has orders already, I would like to define a `HasMany` relation and 

- find all the orders for said customer in a controller method
- create an order for said customer in a controller method

## Acceptance Criteria

- [x] Infer relation metadata from the `hasMany` decorator to construct constrained target repository instance on relational property
- [x] Create a calculated property on the source repository (CustomerRepository in this case) which returns a constrained target repository (for e.g. `customerRepo.<decorated_property_name or relation_name>.update(...)`) using the relation metadata and supplied constraint in the controller method

Code snippet for UX:
```ts
export class CustomerController {
  constructor(
    @repository(CustomerRepository) protected customerRepo: CustomerRepository,
    @repository(OrderRepository) protected orderRepo: OrderRepository
  ) {}

  async function getCustomerOrders(customerId: number): Promise<Order[]> {
    return await customerRepo.orders({id: customerId}).find();
  }

 async function createCustomerOrders(customerId: number, orderData: Partial<Order>): Promise<Order> {
    return await customerRepo.orders({id: customerId}).create(orderData);
  }
  }
```
```ts
export class CustomerRepository {
  // We should be able to inject a factory of OrderRepository by name or instance
  // and infer `Order` from the `OrderRepository`
  @hasMany('orderRepository');  // The argument can be an object to pass in more info
  public readonly orders: (key: Partial<Customer>) => OrderRepository;
  }
```

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

**Comment by raymondfeng** [comment]
```ts
export class CustomerRepository {
  // We should be able to inject a factory of OrderRepository by name or instance
  // and infer `Order` from the `OrderRepository`
  @hasMany('orderRepository');  // The argument can be an object to pass in more info
  public readonly orders: (key: ...) => OrderRepository;
  }
```

```ts
async function getCustomerOrders(customerId: number): Promise<Order[]> {
    return await customerRepo.orders({id: customerId}).find();
}
```

**Comment by b-admike** [comment]
>Does it mean we automatically get the related property customerRepo.orders by:
> add relation in model def of Customer and Order
> instintiate customerRepo and orderRepo in the controller class's constructor?

@jannyHou yeah that's correct.

**Comment by bajtos** [comment]
See also https://github.com/strongloop/loopback-next/issues/1353

**Comment by b-admike** [comment]
Closed as #1438 has landed.
